source/common/compressor/compressor_impl.cc

```diff
@@ -0,0 +1,91 @@
+#include "common/compressor/compressor_impl.h"
+
+#include "envoy/common/exception.h"
+
+#include "common/common/assert.h"
+
+namespace Envoy {
+namespace Compressor {
+
+CompressorImpl::CompressorImpl()
+    : zstream_ptr_(new z_stream(), [](z_stream* z) {
+        deflateEnd(z);
+        delete z;
+      }) {
+  zstream_ptr_->zalloc = Z_NULL;
+  zstream_ptr_->zfree = Z_NULL;
+  zstream_ptr_->opaque = Z_NULL;
+}
+
+CompressorImpl::~CompressorImpl() {}
+
+void CompressorImpl::init(CompressionLevel comp_level, CompressionStrategy comp_strategy,
+                          int window_bits, uint8_t memory_level = 8) {
+  if (initialized_) {
+    throw EnvoyException("Compressor can only be initialized once");
+  }
+
+  const int res = deflateInit2(zstream_ptr_.get(), comp_level, Z_DEFLATED, window_bits,
+                               memory_level, comp_strategy);
+
+  if (res != Z_OK) {
+    // TODO: error mapping
+    throw EnvoyException(fmt::format("Compressor initialization error: '{}'", res));
+  }
+
+  initialized_ = true;
+}
+
+void CompressorImpl::setChunk(uint64_t chunk) { chunk_ = chunk; }
+
+void CompressorImpl::finish(Buffer::Instance& output_buffer) {
+  process(output_buffer, Z_SYNC_FLUSH);
+  commit(output_buffer);
+}
+
+void CompressorImpl::compress(const Buffer::Instance& intput_buffer,
+                              Buffer::Instance& output_buffer) {
+  if (zstream_ptr_->total_out == 0) {
+    reserve(output_buffer);
+  }
+
+  const uint64_t num_slices = intput_buffer.getRawSlices(nullptr, 0);
+  Buffer::RawSlice slices[num_slices];
+  intput_buffer.getRawSlices(slices, num_slices);
+
+  for (const Buffer::RawSlice& input_slice : slices) {
+    zstream_ptr_->avail_in = input_slice.len_;
+    zstream_ptr_->next_in = static_cast<Bytef*>(input_slice.mem_);
+    process(output_buffer, Z_NO_FLUSH);
+  }
+}
```

Comment:
Should there be an explicit flush here once the loop is done?

Reply:
Good question @dnoe 

As I understood from Zlib's documentation (https://www.zlib.net/manual.html)

```... deflate compresses as much data as possible, and stops when the input buffer becomes empty or the output buffer becomes full. It may introduce some output latency (reading input without producing any output) except when forced to flush.

... Forcing flush frequently degrades the compression ratio, so this parameter should be set only when necessary. Some output may be provided even if flush is zero.

... Normally the parameter flush is set to Z_NO_FLUSH, which allows deflate to decide how much data to accumulate before producing output, in order to maximize compression.```

Reply:
What I was thinking of here is that it seems like it might make sense for the `CompressorImpl::finish()` function to be called just before this `compress()` function exits rather than requiring an explicit call to `finish()` which I understand is currently required. I leave this up to you - I think you probably have a better idea about whether this makes sense from the perspective of a user of the Compressor API. If it's normal to make several calls to `compress()` before you're ready to `finish()` then I don't see any reason to add the implicit finish.